### json 불러와서 저장하기

In [1]:
import numpy as np
import pandas as pd
import json
import jpype
from bs4 import BeautifulSoup as bs

data_path = "data"

json_train=[]
for i in range(0,18):
    json_file = json.loads(open("%s/%d.json" % (data_path, i)).read())
    json_train += json_file["articles"]

### json에서 title 형태소 추출

In [2]:
from konlpy.tag import Mecab

mecab = Mecab()

labeled_train = []

for cnt, article in enumerate(json_train):
    text = bs(article["text"],"lxml").text
    
    if cnt % 10000 == 0:
        print cnt

    labeled_train.append({
            "istroll": article["is_troll"],
            "title_morphs": mecab.morphs(article['title'])
        })

labeled_train = pd.DataFrame.from_dict(labeled_train)
labeled_train

/Library/Python/2.7/site-packages/bs4/__init__.py:198: UserWarning: ".." looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.
  '"%s" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.' % markup)
/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://blog.naver.com/kakashi08/110086751501" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://ilwar.com/free/5920
http://ilwar.com/free/6078" looks like a URL. Beautiful Soup is not an HTTP client. You should probabl

0
10000

/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://image.kilho.net/?pk=1443527



Free" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://www.youtube.com/watch?v=ByhP478ZPQI

ㄹㅇ존좋" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Python/2.7/site-packages/bs4/


20000

/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://imgur.com/rBtm1LI.jpg" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://m.ilwar.com/poli/17062" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "ht


30000

/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://www.ilbe.com/2691808050" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://withgonggam.tistory.com/m/post/view/id/1316" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)



40000

/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://m.khan.co.kr/view.html?category=2&med_id=khan&artid=201401071905352&code=361102" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://ilwar.com/free/44658" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Python


50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000


/Library/Python/2.7/site-packages/bs4/__init__.py:207: UserWarning: "http://www.ohmynews.com/NWS_Web/View/at_pg.aspx?CNTN_CD=A0001947743" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)


istroll                                       title_morphs
0        False                             [일간, 워스트, 마을, 회관, 입니다]
1        False                               [일간, 워스트, 일, 워, 입니다]
2        False                                             [오, !]
3        False                                             [1, 등]
4        False                                 [여기, 가, 수용소, 냐, ?]
5        False                                   [여기, 가, 일, 워, !]
6        False                                             [지, 젼]
7        False                       [함수, 니, +, 빛돌, 이, 합동, 콘, 해라]
8        False  [캬, ., .., 미루, 고, 미루, 다, 보, 면, 역시, rainy, 님, 이...
9        False                        [일간, ㅋㅋㅋㅋㅋㅋㅋ워슼ㅋㅋㅋㅋㅋ틐ㅋㅋㅋㅋㅋㅋ]
10       False                                    [순위, 놀이, ㅁ, ㅇㅎ]
11       False                                 [민영화, 말, 고, 사영, 화]
12       False  [어떻게, 해야, 가장, 뛰어난, 첫, 번, 째, 게시, 글, 이, 될지, 고민, ...
13       False                                     [경축, 세, 가지, !]
14       False                              [덕, 후판, 만들, 어, 주, 세요]
15       False             [민영화, 누르, 면, 비, 추천, 했, 습니다, 라고, 뜨, 네여]
16       False                                              [엉엉엉]
17       False                              [기억, 이, 나, 지, 않, 습니다]
18       False                                        [안녕, 하, 세요]
19       False                                [아이디, 는, 언제, 쓰, 나요]
20       False               [ㅇㅂ는, 애국, 보수, 면, 여긴, 머, 라, 해야, 하, 지]
21       False                                 [어휴, 워, 충이, 들, ㅉㅉ]
22       False                        [일, 베, 편의점, 담배, 호빵, 사건, 전말]
23       False              [오픈, 기념, 으로, 일, 워, 로고, 만들, 어, 봤, 습니다]
24       False                                          [아야나미, !]
25       False                          [철도, 민영화, 한방, 에, 정리, 해준다]
26       False                                 [박정근, 만세, 해, 봐, .]
27       False                [터키, 노조, 프랑스, 노조, 는, 그럼, 종북, 좌파, 냐]
28       False                  [모바일, 버전, 두, 얼릉, 오픈, 하, 길, ., ..]
29       False                                             [유루유리]
...        ...                                                ...
172014   False                           [휴, ., .., 우울, 하, 더라, 농]
172015    True                           [탈모, 에, 대해서, 질문, 입니다, .]
172016   False                                 [뜻밖, 의, 불꽃, 싸, 다구]
172017   False               [아, 이마스, 보, 는데, 미키, 가, 잘, 나왔, 네요, .]
172018   False                [지금, 모바일, 사진, 지원, 안, 되, 시, 는, 분, ?]
172019   False                           [어, ?, 저, 트롤, 밭, 인가요, ?]
172020   False                                           [벌레, 출현]
172021   False                              [저, 왜, 트롤, 인가요, ., .]
172022   False                            [이명박, 의, 독특, 한, 영어, 인사]
172023   False                  [농부, 들, 한테, 궁금, 한, 게, 있, 다, 농, .]
172024   False  [본, 농, 은, 될, 수, 있, 음, 사람, 의, 장점, 을, 보, 려고, 노력,...
172025   False                             [트롤, 밭, ., ., 무서웠, 어요]
172026   False                            [PS, 4, 비둘기, 연애, 시뮬레이션]
172027   False             [주연, 테크, 서, 나온, 제이탭, s, 2, 를, 샀, 다, 농]
172028   False                                    [최근, 얼굴, 책, 에서]
172029   False                                    [2014, 년, ., .]
172030   False                                [그러, 고, 보, 니, ., .]
172031   False                  [농, 농, 들, 마우스, 추천, 해, 달, 라, 농, !]
172032   False          [정규직, 과, 비정규직, 이, 함께, 한, cnm, 노사, 잠정, 합의]
172033   False                               [어디, 서, 본, 듯, 한데, 2]
172034   False  [(, 속보, ), 땅콩, 부사장, 님, 에게, 구속, 영장, 이, 발부, 되, 었...
172035   False                                 [농약, 충전, 용, ., ..]
172036   False                           [umc, ,, 이용, (, 뻘, 글, )]
172037   False   [안녕, 하, 세요, ~, 일, 워, 개, 장기, 보, 고, 가입, 했, 습니다, .]
172038   False     [토토, 가, 열풍, ,, 90, 년, 대, 음악, 은, 정말, 최고, 일까, ?]
172039   False                              [몸, 이, 약해진, 건지, ., .]
172040   False          [살, 면서, 좋, 은, 집, 주인, 은, 본, 적, 이, 없, 구, 농]
172041   False                 [C, 9, 외국, 섭, 하, 시, 는, 분, 계신가요, .]
172042   False         [끄, 아, 연말, 이, 라니, ., ..,

###bag of words

In [3]:
from sklearn.feature_extraction.text import CountVectorizer

max_features = 1000

vectorizer = CountVectorizer(analyzer = "word", tokenizer = None, preprocessor = None,
                            stop_words = None, max_features=max_features)

train_data_features = vectorizer.fit_transform((labeled_train["title_morphs"].apply(lambda x:" ".join(x)))).toarray()

col = ["bow_title_%s" % data for data in vectorizer.get_feature_names()]
labeled_train = pd.concat([labeled_train, pd.DataFrame(train_data_features, columns = col, index=labeled_train.index)],axis=1)

labeled_train

### predictor, model 생성

In [4]:
from sklearn.ensemble import RandomForestClassifier

label = 'istroll'
pre = labeled_train.columns.drop(['title_morphs', label])

model = RandomForestClassifier(n_estimators=10, n_jobs=6)

###cross validation

In [5]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, labeled_train[pre], labeled_train[label], cv=cv_value, scoring="roc_auc")
cv_result = scores.mean()

print(cv_result)

0.612752193303


##Result
  * title의 morphs를 bag of words로 변환 (feature : 1,000개) - 0.612752193303

In [7]:
import pickle

pickle.dump(labeled_train[pre], open("title_%d.p" % max_features, "wb"), protocol = pickle.HIGHEST_PROTOCOL)

bow_title_017  bow_title_10  bow_title_100  bow_title_11  \
0                   0             0              0             0   
1                   0             0              0             0   
2                   0             0              0             0   
3                   0             0              0             0   
4                   0             0              0             0   
5                   0             0              0             0   
6                   0             0              0             0   
7                   0             0              0             0   
8                   0             0              0             0   
9                   0             0              0             0   
10                  0             0              0             0   
11                  0             0              0             0   
12                  0             0              0             0   
13                  0             0              0             0   
14                  0             0              0             0   
15                  0             0              0             0   
16                  0             0              0             0   
17                  0             0              0             0   
18                  0             0              0             0   
19                  0             0              0             0   
20                  0             0              0             0   
21                  0             0              0             0   
22                  0             0              0             0   
23                  0             0              0             0   
24                  0             0              0             0   
25                  0             0              0             0   
26                  0             0              0             0   
27                  0             0              0             0   
28                  0             0              0             0   
29                  0             0              0             0   
...               ...           ...            ...           ...   
172014              0             0              0             0   
172015              0             0              0             0   
172016              0             0              0             0   
172017              0             0              0             0   
172018              0             0              0             0   
172019              0             0              0             0   
172020              0             0              0             0   
172021              0             0              0             0   
172022              0             0              0             0   
172023              0             0              0             0   
172024              0             0              0             0   
172025              0             0              0             0   
172026              0             0              0             0   
172027              0             0              0             0   
172028              0             0              0             0   
172029              0             0              0             0   
172030              0             0              0             0   
172031              0             0              0             0   
172032              0             0              0             0   
172033              0             0              0             0   
172034              0             0              0             0   
172035              0             0              0             0   
172036              0             0              0             0   
172037              0             0              0             0   
172038              0             0              0             0   
172039              0             0              0             0   
172040              0             0              0             0